common/crypto_utils/avalon_crypto_utils/crypto_utility.py

```diff
@@ -108,14 +123,13 @@ def decrypt_data(encryption_key, data, iv=None):
     if not data:
         logger.debug("Outdata is empty, nothing to decrypt")
         return data
-    data_byte = crypto.base64_to_byte_array(data)
-    logger.debug("encryption_key: %s", encryption_key)
-    if iv is not None:
-        decrypt_result = crypto.SKENC_DecryptMessage(encryption_key,
-                                                     iv, data_byte)
-    else:
-        decrypt_result = crypto.SKENC_DecryptMessage(encryption_key, data_byte)
-    result = crypto.byte_array_to_string(decrypt_result)
+
+    data_byte = base64_to_byte_array(data)
+    cipher = AES.new(bytes(encryption_key), AES.MODE_GCM, bytes(iv))
+    # Split data into tag and ciphered data
+    ciphered_data = bytes(data_byte[:(len(data_byte)-TAG_SIZE)])
+    tag = bytes(data_byte[-1*TAG_SIZE:])
+    result = cipher.decrypt_and_verify(ciphered_data, tag).decode("utf-8")
```

Comment:
if iv is None the it's assumed that 12 bytes iv is prepended in encrypted data. decrypt_data() should consider this.

Reply:
Below is code snippet.
```
def decrypt_data(self, enc_data_bytes, session_key, iv=None):
        # if iv is None the it's assumed that 12 bytes iv is 
        # prepended in encrypted data
        if iv is None:
            iv_length = 12
            iv = enc_data_bytes[:iv_length]           
        else:
            iv_length = len(iv)

        cipher_aes = AES.new(session_key, AES.MODE_GCM, iv)
        tag_length = 16
        ciphertext_len = len(enc_data_bytes) - iv_length - tag_length
        ciphertext = enc_data_bytes[iv_length: iv_length + ciphertext_len]
        tag = enc_data_bytes[-tag_length:]

        result = cipher_aes.decrypt_and_verify(ciphertext, tag)
        result = result.decode("utf-8")
        return result
```

Reply:
Updated. Please review.